File: 2692-take-gifts-from-the-richest-pile/solution.cpp
class Solution {
public:
    long long pickGifts(vector<int>& gifts, int k) {
        priority_queue<int> h(gifts.begin(), gifts.end());

        for (int i = 0; i < k; i++) {
            int maxElement = h.top();
            h.pop();

            h.push(sqrt(maxElement)); // heapify
        }

        long long res = 0;
        while (!h.empty()) {
            res += h.top();
            h.pop();
        }

        return res;
    }
};

File: 2692-take-gifts-from-the-richest-pile/solution.py
class Solution:
    def pickGifts(self, gifts: List[int], k: int) -> int:
        # py has min heap so use - minus sign to simulate max heap
        h = [-g for g in gifts]
        heapq.heapify(h)

        for _ in range(k):
            maxElement = -heapq.heappop(h)

            heapq.heappush(h, -math.floor(math.sqrt(maxElement)))
        
        res = 0
        while h:
            # now use - to make it back positive
            res -= heapq.heappop(h)
        
        return res
